JPA Map Storage with CRDB: ConcurrentLoginTest failures

### Describe the bug

**org.keycloak.testsuite.admin.concurrency.ConcurrentLoginTest**  presents failures/errors with `CockroachDB`:

```
[ERROR]   ConcurrentLoginTest.concurrentLoginSingleUserSingleClient:170->AbstractConcurrencyTest.run:63->AbstractConcurrencyTest.run:122 » Runtime There were failures in threads. Failures count: 2
```

- Required JPA stores to reproduce the failure: **authenticationSessions**

```
[ERROR]   ConcurrentLoginTest.concurrentCodeReuseShouldFail:249->AbstractConcurrencyTest.run:63->AbstractConcurrencyTest.run:122 » Runtime There were failures in threads. Failures count: 2
```

- Required JPA stores to reproduce the failure: **userSessions**

For both tests, logs show the following error:

```
Caused by: org.postgresql.util.PSQLException: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_WRITE_TOO_OLD - WriteTooOld flag converted to WriteTooOldError): "sql txn" meta={id=7af23746 key=/Table/3142/1/"\x9cX4\x9e\xaf\x94F/\xac\xe41~H@\"\x10"/0 pri=0.05982192 epo=0 ts=1658239825.904432104,1 min=1658239825.904379733,0 seq=2} lock=true stat=PENDING rts=1658239825.904379733,0 wto=false gul=1658239826.404379733,0
  Hint: See: https://www.cockroachlabs.com/docs/v21.2/transaction-retry-error-reference.html#retry_write_too_old
```

### Version

main

### Expected behavior

Tests should run without failures or errors



### Actual behavior

_No response_

### How to Reproduce?

1- Setup CockroachDB according to the instructions in https://github.com/keycloak/keycloak/discussions/9596
2- Run the test with the correct parameters, enabling JPA storage for **authenticationSessions**/**userSessions** only:

```
cd testsuite/integration-arquillian

mvn clean install -Pmap-storage -Dpageload.timeout=90000 -Dkeycloak.authSession.map.storage.provider=jpa \
-Dkeycloak.userSession.map.storage.provider=jpa \
-Dkeycloak.map.storage.connectionsJpa.url=jdbc:postgresql://localhost:26257/keycloak \
-Dkeycloak.map.storage.connectionsJpa.user=keycloak -Dkeycloak.map.storage.connectionsJpa.password=pass \
-Dtest=ConcurrentLoginTest
```

### Anything else?

_No response_

--- a/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java
+++ b/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java
@@ -31,6 +31,7 @@ import org.keycloak.common.ClientConnection;
 import org.keycloak.common.Profile;
 import org.keycloak.common.constants.ServiceAccountConstants;
 import org.keycloak.common.util.KeycloakUriBuilder;
+import org.keycloak.common.util.Resteasy;
 import org.keycloak.constants.AdapterConstants;
 import org.keycloak.events.Details;
 import org.keycloak.events.Errors;
@@ -48,6 +49,7 @@ import org.keycloak.models.RealmModel;
 import org.keycloak.models.UserModel;
 import org.keycloak.models.UserSessionModel;
 import org.keycloak.models.utils.AuthenticationFlowResolver;
+import org.keycloak.models.utils.KeycloakModelUtils;
 import org.keycloak.protocol.oidc.OIDCAdvancedConfigWrapper;
 import org.keycloak.protocol.oidc.OIDCLoginProtocol;
 import org.keycloak.protocol.oidc.TokenExchangeContext;
@@ -105,6 +107,7 @@ import javax.ws.rs.InternalServerErrorException;
 import javax.ws.rs.OPTIONS;
 import javax.ws.rs.POST;
 import javax.ws.rs.Path;
+import javax.ws.rs.WebApplicationException;
 import javax.ws.rs.core.Context;
 import javax.ws.rs.core.HttpHeaders;
 import javax.ws.rs.core.MediaType;
@@ -172,6 +175,32 @@ public class TokenEndpoint {
     @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
     @POST
     public Response processGrantRequest() {
+        // grant request needs to be run in a retriable transaction as concurrent execution of this action can lead to
+        // exceptions on DBs with SERIALIZABLE isolation level.
+        Object result = KeycloakModelUtils.runJobInRetriableTransaction(this.session.getKeycloakSessionFactory(), kcSession -> {
+            try {
+                RealmModel realmModel = kcSession.realms().getRealm(realm.getId());
+                kcSession.getContext().setRealm(realmModel);
+                // create another instance of the endpoint that will be run within the new session.
+                Resteasy.pushContext(KeycloakSession.class, kcSession);
+                TokenEndpoint other = new TokenEndpoint(new TokenManager(), realmModel, new EventBuilder(realmModel, kcSession, clientConnection));
+                ResteasyProviderFactory.getInstance().injectProperties(other);
+                return other.processGrantRequestInternal();
+            } catch (WebApplicationException we) {
+                // WebApplicationException needs to be returned and treated (rethrown) by the calling code because the new transaction
+                // still needs to be committed when this exception is thrown. It captures final business states that won't change when
+                // being retried, like an invalid code.
+                return we;
+            }
+        }, 10, 100);
+        if (WebApplicationException.class.isInstance(result)) {
+            throw (WebApplicationException) result;
+        } else {
+            return (Response) result;
+        }
+    }
+
+    private Response processGrantRequestInternal() {
         cors = Cors.add(request).auth().allowedMethods("POST").auth().exposedHeaders(Cors.ACCESS_CONTROL_ALLOW_METHODS);
 
         MultivaluedMap<String, String> formParameters = request.getDecodedFormParameters();
@@ -179,7 +208,7 @@ public class TokenEndpoint {
         if (formParameters == null) {
             formParameters = new MultivaluedHashMap<>();
         }
-        
+
         formParams = formParameters;
         grantType = formParams.getFirst(OIDCLoginProtocol.GRANT_TYPE_PARAM);
 

